File: ts/test/entity/convert/ConvertEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { ForeignExchangeRatesSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('ConvertEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when FOREIGNEXCHANGERATES_TEST_LIVE=TRUE.
  afterEach(liveDelay('FOREIGNEXCHANGERATES_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = ForeignExchangeRatesSDK.test()
    const ent = testsdk.Convert()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.FOREIGN_EXCHANGE_RATES_TEST_LIVE
    for (const op of ['create', 'list']) {
      if (maybeSkipControl(t, 'entityOp', 'convert.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const convert_ref01_ent = client.Convert()
    let convert_ref01_data = setup.data.new.convert['convert_ref01']
    convert_ref01_data['amount'] = setup.idmap['amount01']
    convert_ref01_data['from'] = setup.idmap['from01']
    convert_ref01_data['to'] = setup.idmap['to01']

    convert_ref01_data = await convert_ref01_ent.create(convert_ref01_data)
    assert(null != convert_ref01_data)


    // LIST
    const convert_ref01_match: any = {}
    convert_ref01_match['amount'] = setup.idmap['amount01']
    convert_ref01_match['from'] = setup.idmap['from01']
    convert_ref01_match['to'] = setup.idmap['to01']

    const convert_ref01_list = await convert_ref01_ent.list(convert_ref01_match)

    assert(!isempty(select(convert_ref01_list, { id: convert_ref01_data.id })))


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/convert/ConvertTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = ForeignExchangeRatesSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['convert01','convert02','convert03','convert01','convert02','convert03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID': idmap,
    'FOREIGN_EXCHANGE_RATES_TEST_LIVE': 'FALSE',
    'FOREIGN_EXCHANGE_RATES_TEST_EXPLAIN': 'FALSE',
    'FOREIGN_EXCHANGE_RATES_APIKEY': 'NONE',
  })

  idmap = env['FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID']

  const live = 'TRUE' === env.FOREIGN_EXCHANGE_RATES_TEST_LIVE

  if (live) {
    client = new ForeignExchangeRatesSDK(merge([
      {
        apikey: env.FOREIGN_EXCHANGE_RATES_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.FOREIGN_EXCHANGE_RATES_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  

File: go/test/convert_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/foreign-exchange-rates-sdk/go"
	"github.com/voxgig-sdk/foreign-exchange-rates-sdk/go/core"

	vs "github.com/voxgig-sdk/foreign-exchange-rates-sdk/go/utility/struct"
)

func TestConvertEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Convert(nil)
		if ent == nil {
			t.Fatal("expected non-nil ConvertEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := convertBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "convert." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		convertRef01Ent := client.Convert(nil)
		convertRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "convert"}), "convert_ref01"))
		convertRef01Data["from"] = setup.idmap["from01"]
		convertRef01Data["to"] = setup.idmap["to01"]

		convertRef01DataResult, err := convertRef01Ent.Create(convertRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		convertRef01Data = core.ToMapAny(entityData(convertRef01DataResult))
		if convertRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if convertRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LOAD
		convertRef01MatchDt0 := map[string]any{
			"id": convertRef01Data["id"],
		}
		convertRef01DataDt0Loaded, err := convertRef01Ent.Load(convertRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		convertRef01DataDt0LoadResult := core.ToMapAny(entityData(convertRef01DataDt0Loaded))
		if convertRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if convertRef01DataDt0LoadResult["id"] != convertRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func convertBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "convert", "ConvertTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read convert test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse convert test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"convert01", "convert02", "convert03", "from01", "to01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID": idmap,
		"FOREIGN_EXCHANGE_RATES_TEST_LIVE":      "FALSE",
		"FOREIGN_EXCHANGE_RATES_TEST_EXPLAIN":   "FALSE",
		"FOREIGN_EXCHANGE_RATES_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["FOREIGN_EXCHANGE_RATES_TEST_CONVERT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FOREIGN_EXCHANGE_RATES_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["FOREIGN_EXCHANGE_RATES_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewForeignExchangeRatesSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FOREIGN_EXCHANGE_RATES_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FOREIGN_EXCHANGE_RATES_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
